packages/core/admin/admin/src/content-manager/components/DynamicZone/DynamicComponent.tsx

```diff
@@ -109,10 +109,26 @@ const DynamicComponent = ({
         displayedValue: `${friendlyName}${mainValue}`,
         icon,
       },
-      onMoveItem: onMoveComponent,
-      onGrabItem,
-      onDropItem,
-      onCancel,
+      onMoveItem(newIndex: number | Array<number>, currentIndex: number | Array<number>) {
+        if (typeof newIndex === 'number' && typeof currentIndex === 'number') {
+          onMoveComponent(newIndex, currentIndex);
+        }
+      },
+      onDropItem(currentIndex: number | Array<number>) {
+        if (typeof currentIndex === 'number' && onDropItem) {
+          onDropItem(currentIndex);
+        }
+      },
+      onGrabItem(index: number | Array<number>) {
+        if (typeof index === 'number' && onGrabItem) {
+          onGrabItem(index);
+        }
+      },
+      onCancel(index: number | Array<number>) {
+        if (typeof index === 'number' && onCancel) {
+          onCancel(index);
+        }
+      },
```

Comment:
this is quite messy. Can the types for Drag&Drop infer which index type you've passed and do the rest? It's not a great API if we need to do this.

Reply:
Even I disliked doing this, but for callbacks there is no other option to override union of parameters, type check fails.
As you suggested I don't think we can make it generic as we need to handle each passed type onMove() inside DnD hook. Can you please check on your end and then we can have a call.